### PR TITLE
2020 passthrough asserts

### PR DESCRIPTION
**Pull request by drowe67**
Addressing freedv-gui Issue: https://github.com/drowe67/freedv-gui/issues/106

1. Fix corner case where number of 8kHz modem samples, when resampled to 16 kHz, overflowed speech output buffer and fired assert.
2. Removed unneeded assert in fdmdv 16<->8k resamplers that restricted operation to an even number of samples in the 8k buffer.  The resamplers actually work fine with odd numbers of samples in the 8k buffers, e.g. 159 samples at 8k -> 318 at 16k -> 159 at 8k.
1. Wrote ctests for 16<->8k resamplers.

@tmiw :point_up:

### CODE REVIEW

**Comment by tmiw** [comment]
@drowe67, 88/89 unit tests passed on macOS, albeit with a few crashes from octave-gui. The failing case below might also be an Octave issue but isn't related to these changes.

```44/89 Test #44: test_freedv_api_700D_burble ....................***Failed  Required regular expression not found. Regex=[output speech samples: 0
]  0.98 sec```

But yeah, I'm fine with this PR. 👍

**Comment by tmiw** [comment]
BTW, full test output from the failing test:

```
HT-TM05:build_osx mooneer$ ctest -V -R test_freedv_api_700D_burble
UpdateCTestConfiguration  from :/Users/mooneer/devel/freedv-gui/codec2/build_osx/DartConfiguration.tcl
Parse Config file:/Users/mooneer/devel/freedv-gui/codec2/build_osx/DartConfiguration.tcl
UpdateCTestConfiguration  from :/Users/mooneer/devel/freedv-gui/codec2/build_osx/DartConfiguration.tcl
Parse Config file:/Users/mooneer/devel/freedv-gui/codec2/build_osx/DartConfiguration.tcl
Test project /Users/mooneer/devel/freedv-gui/codec2/build_osx
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 44
    Start 44: test_freedv_api_700D_burble

44: Test command: /bin/sh "-c" "cd /Users/mooneer/devel/freedv-gui/codec2/build_osx/src;
                            ./freedv_tx 700D ../../raw/ve9qrp.raw - |
                            ./cohpsk_ch - - -8 --Fs 8000 |
                            ./freedv_rx 700D - /dev/null --squelch -2"
44: Test timeout computed to be: 1500
44: cohpsk_ch: Fs: 8000 NodB: -8.00 foff: 0.00 Hz fading: 0 nhfdelay: 0 clip: 32767.00 ssbfilt: 1 complexout: 0
44: cohpsk_ch: SNR3k(dB):    -9.80  C/No....:    24.98
44: cohpsk_ch: peak.....: 17030.04  RMS.....:  7056.30   CPAPR.....:  7.65 
44: cohpsk_ch: Nsamples.:   898560  clipped.:     0.00%  OutClipped:  2.34%
44: cohpsk_ch: WARNING output clipping
44: frames decoded: 745  output speech samples: 1280
1/1 Test #44: test_freedv_api_700D_burble ......***Failed  Required regular expression not found. Regex=[output speech samples: 0
]  0.84 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   0.85 sec

The following tests FAILED:
	 44 - test_freedv_api_700D_burble (Failed)
Errors while running CTest
```

**Comment by drowe67** [comment]
@tmiw that ctest fail is in a similar area of the code modified in this PR - I'll look into it ...

At some point it would be good to sort out any ctests failing on macOS just in case the fails are pointing at some more corner cases ...

**Comment by tmiw** [comment]
@drowe67, looks like it's the same issue that occurred with Octave a while ago. Changing CMakeLists.txt as follows causes all the tests to at least run without segfaulting (but test_freedv_api_700D_burble still fails):

```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 2fb06d63..b2a12f6c 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -349,7 +349,7 @@ if(UNITTEST)
     add_test(NAME test_fdmdv_16to8
              COMMAND sh -c "cd ${CMAKE_CURRENT_SOURCE_DIR}/octave;
                             ${CMAKE_CURRENT_BINARY_DIR}/unittest/t16_8;
-                            DISPLAY=\"\" echo \"diff_fft_mag('in8.raw','out8.raw'); quit;\" | octave --no-gui -qf
+                            DISPLAY=\"\" echo \"diff_fft_mag('in8.raw','out8.raw'); quit;\" | octave-cli -qf
                             ")
              set_tests_properties(test_fdmdv_16to8 PROPERTIES PASS_REGULAR_EXPRESSION "PASS")
 
@@ -357,12 +357,12 @@ if(UNITTEST)
     add_test(NAME test_fdmdv_16to8_short
              COMMAND sh -c "cd ${CMAKE_CURRENT_SOURCE_DIR}/octave;
                             ${CMAKE_CURRENT_BINARY_DIR}/unittest/t16_8_short;
-                            DISPLAY=\"\" echo \"diff_fft_mag('in8_short.raw','out8_short.raw'); quit;\" | octave --no-gui -qf
+                            DISPLAY=\"\" echo \"diff_fft_mag('in8_short.raw','out8_short.raw'); quit;\" | octave-cli -qf
                             ")
              set_tests_properties(test_fdmdv_16to8_short PROPERTIES PASS_REGULAR_EXPRESSION "PASS")
 
     add_test(NAME test_CML_ldpcut
-             COMMAND sh -c "cd ${CMAKE_CURRENT_SOURCE_DIR}/octave; SHORT_VERSION_FOR_CTEST=1 octave --no-gui -qf ldpcut.m")
+             COMMAND sh -c "cd ${CMAKE_CURRENT_SOURCE_DIR}/octave; SHORT_VERSION_FOR_CTEST=1 octave-cli -qf ldpcut.m")
              set_tests_properties(test_CML_ldpcut PROPERTIES PASS_REGULAR_EXPRESSION "Nerr: 0")
 
     # check channel simulator measures correct Peak to Average Power Ratio (about 0dB) with a sine wave input signal
@@ -422,7 +422,7 @@ if(UNITTEST)
     add_test(NAME test_OFDM_modem_octave_qam16_uncoded
              COMMAND sh -c "cd ${CMAKE_CURRENT_SOURCE_DIR}/octave;
                             echo \"ofdm_tx('test_qam16.raw','qam16c1',3,12); ofdm_rx('test_qam16.raw','qam16c1', 0.05); quit\" |
-                            DISPLAY=\"\" octave --no-gui")
+                            DISPLAY=\"\" octave-cli")
              set_tests_properties(test_OFDM_modem_octave_qam16_uncoded PROPERTIES PASS_REGULAR_EXPRESSION "Pass")
 
     # noise free uncoded 700D test, including reading and writing payload bits
@@ -843,12 +843,12 @@ endif(NOT APPLE)
 
      # Octave FSK Modem, to make sure we don't break reference simulation
      add_test(NAME test_fsk_lib
-             COMMAND sh -c "cd ${CMAKE_CURRENT_SOURCE_DIR}/octave; DISPLAY=\"\" octave --no-gui -qf fsk_lib_demo.m")
+             COMMAND sh -c "cd ${CMAKE_CURRENT_SOURCE_DIR}/octave; DISPLAY=\"\" octave-cli -qf fsk_lib_demo.m")
              set_tests_properties(test_fsk_lib PROPERTIES PASS_REGULAR_EXPRESSION "PASS")
 
      add_test(NAME test_fsk_modem_octave_port
              COMMAND sh -c "cd ${CMAKE_CURRENT_SOURCE_DIR}/octave;
-                            PATH_TO_TFSK=${CMAKE_CURRENT_BINARY_DIR}/unittest/tfsk octave --no-gui -qf tfsk.m")
+                            PATH_TO_TFSK=${CMAKE_CURRENT_BINARY_DIR}/unittest/tfsk octave-cli -qf tfsk.m")
              set_tests_properties(test_fsk_modem_octave_port PROPERTIES PASS_REGULAR_EXPRESSION "PASS")
 
      add_test(NAME test_fsk_modem_mod_demod
@@ -894,7 +894,7 @@ endif(NOT APPLE)
 
      # Octave 4FSK LLR reference simulation - make sure this keeps working
      add_test(NAME test_fsk_lib_4fsk_ldpc
-             COMMAND sh -c "cd ${CMAKE_CURRENT_SOURCE_DIR}/octave; DISPLAY=\"\" octave --no-gui -qf fsk_lib_ldpc_demo.m")
+             COMMAND sh -c "cd ${CMAKE_CURRENT_SOURCE_DIR}/octave; DISPLAY=\"\" octave-cli -qf fsk_lib_ldpc_demo.m")
              set_tests_properties(test_fsk_lib_4fsk_ldpc PROPERTIES PASS_REGULAR_EXPRESSION "PASS")
 
      # Command line Unique Word (UW) framer in hard decision mode
```

**Comment by drowe67** [comment]
Thanks @tmiw I've moved to `octave-cli` and added a higher verbosity level for that burble test.  Pls email me the log you see when it fails.

The idea of that test is 700D should not break squelch due to random syncs on noise.  You machine has different initial conditions (e.g. random number generator) so we're getting a different result - it's letting one frame of audio through on the 60 second test run.

**Comment by tmiw** [comment]
@drowe67, just emailed the results with the additional output. 👍

**Comment by drowe67** [comment]
So frame 193 is where it let the one frame of decoded speech out:
```
192 nin: 1280 st: search euw:  4  0 mf:  0 f:  -4.5 pbw: 1 snr: -1.8 eraw:   0 ecdd:   0 iter:   0 pcc:   0 rxst: ----
193 nin: 1280 st: trial  euw:  4  1 mf:  0 f:  -4.3 pbw: 1 snr: -2.2 eraw:   0 ecdd:   0 iter:  60 pcc: 112 rxst: -BST
194 nin: 1280 st: trial  euw:  5  2 mf:  0 f:  -4.0 pbw: 1 snr: -2.8 eraw:   0 ecdd:   0 iter: 100 pcc:  44 rxst: EBST
```
For some reason (probably just bad luck as it's all random data at -8dB SNR) all 112 parity checks (`pcc` field) were met for the LDPC decoder, which is how I'm testing if we should stay squelched to prevent the burbling.  This is why CRCs are a good idea I guess, but we don't have the spare bits for 700D.

I might leave this one for now, and ponder it later.  That was on frame out of a 120 second sample, so still not too bad.
